### ISSUE TITLE
coo_matrix file

### ISSUE THREAD

**Issue by rocketeer1998** [acycliq/pciSeq#4]
Hi @acycliq . Thanks for your great work on pciSeq! I'd like to know how to get the labeled image as a coo_matrix file? Which software is usually used?

**Comment by acycliq** [comment]
Hi there,

Thanks for your interest. Assuming that the label image is the 2D array returned by your segmentation you can do:

```
from scipy.sparse import coo_matrix

coo = coo_matrix(label_image)

```

There are might be cases (not very often) where your segmentation spits out a label image where the labels arent a sequence of consecutive integers (ie without any gaps). Or they are alphanumeric strings  
Hence, I would say, first relabel your label image (see function below) and then convert it to a sparse matrix

```
def reorder_labels(label_image):
    """
    rearranges the labels so that they are a sequence of integers
    """
    _, idx = np.unique(label_image.flatten(), return_inverse=True)
    out = idx.reshape(label_image.shape)
    return out
```

To recap, the full operation would be something like

```
from scipy.sparse import coo_matrix

label_img =  reorder_labels(label_image)
coo = coo_matrix(label_img)
```

Then use that `coo` as your input to pciSeq. To be fair, I need to include that `reorder_labels` function into my code so that it is getting called if needed without the end-user having to do it on his own, beforehand.

Let me know if you have any problems please

**Comment by rocketeer1998** [comment]
> Hi there,
> 
> Thanks for your interest. Assuming that the label image is the 2D array returned by your segmentation you can do:
> 
> ```
> from scipy.sparse import coo_matrix
> 
> coo = coo_matrix(label_image)
> ```
> 
> There are might be cases (not very often) where your segmentation spits out a label image where the labels arent a sequence of consecutive integers (ie without any gaps). Or they are alphanumeric strings Hence, I would say, first relabel your label image (see function below) and then convert it to a sparse matrix
> 
> ```
> def reorder_labels(label_image):
>     """
>     rearranges the labels so that they are a sequence of integers
>     """
>     _, idx = np.unique(label_image.flatten(), return_inverse=True)
>     out = idx.reshape(label_image.shape)
>     return out
> ```
> 
> To recap, the full operation would be something like
> 
> ```
> from scipy.sparse import coo_matrix
> 
> label_img =  reorder_labels(label_image)
> coo = coo_matrix(label_img)
> ```
> 
> Then use that `coo` as your input to pciSeq. To be fair, I need to include that `reorder_labels` function into my code so that it is getting called if needed without the end-user having to do it on his own, beforehand.
> 
> Let me know if you have any problems please

Thanks for your quick reply! I'll try it out today and go back to you later. In addition, why should we use scRNA-seq marker expression matrix as one of the pciSeq inputs? Finally, can I visualize my result in URL like https://acycliq.github.io/ca1/ ?

**Comment by acycliq** [comment]
you need single cell data because this is where the cell types come from. The cell calling algorithm assigns spots to cells while at the same time, tries to find which cell type matches the best possible way the pattern that emerges from these assignments

Yes, you can have the viewer work with your own data, let me know when you get there....
Just remember to run pciSeq with the `save_data` option set to `True`, something like:

```
opts={ 'save_data': True}
res = pciSeq.fit(spots_df, label_image, scRNA_df, opts)
```

That should save the output at `tsv` files to your temp dir. The files are needed for the viewer

**Comment by rocketeer1998** [comment]
I've got all the output files, including `cellBoundaries.tsv`, `cellData.tsv` and `geneData.tsv`. How can I proceed to visualize these results in URL?

**Comment by acycliq** [comment]
Hi there

Had to do a few changes...
Can you follow the readme of this repo `https://github.com/acycliq/visage`  and tell me if it works for you?

**Comment by acycliq** [comment]
Thats weird, Since you are fetching the original demo data i would say it has to do with your [config.js](https://github.com/acycliq/visage/blob/main/viewer/js/config.js) file. Have you saved your data under `ROOT\viewer\data`, just replace the three demo tsvs with yours. 
Let me know how it goes, if your project is on some public repo, I can have a look too

**Comment by acycliq** [comment]
ok, I am assuming then you are trying to open in your browser `E:\Downloads\visage-main\visage-main\index.html`. In other words you `cd` into your `E:\Downloads\visage-main\visage-main` and then you create your local server.
Can you try using these in your `config.js` please
```
cellData: [{mediaLink: './data/cellData.tsv', size: "21267078"}],
geneData: [{mediaLink: './data/geneData.tsv' size: "74721272"}],
cellBoundaries: [{mediaLink: './data/cellBoundaries.tsv', size: "10394177"}],
```

**Comment by rocketeer1998** [comment]
Thanks for your patience! And I'd like to fix this problem today. I've edited my `config.js` file. However, the problem still exists. In addition, I find that no matter where I `cd` to, I can always create a local server that is related to your pciSeq data. Why?

**Comment by acycliq** [comment]
Can you make a public repo so I can have a look?

**Comment by acycliq** [comment]
Can you share the javascript code by any chance, the viewer code that loads up the demo data instead of yours?

**Comment by rocketeer1998** [comment]
I didn't use javascript. I visualized my data following your instruction here,
https://github.com/acycliq/visage
which means I used totally the same scripts as in this repo, except for the configuration files. As for some configuration parameters, such as roi, for now it's ok for you to use the default value. 

The size of my tsv files are as follows,
1.cellBoundaries.tsv: 4,823,294 bytes
2.cellData.tsv: 5,059,483 bytes
3.geneData.tsv: 90,705,878 bytes

Two questions:
1. Why should we declare the size of tsv files in `config.js`?
2. Is it possible to do cell-type annotation without single-cell RNA data? In my opinion, we can speculate on cell type easily by calculating the proportion of each marker gene in each cell.

**Comment by acycliq** [comment]
You need to clone/download the viewer code (ie the visage repo), do the edits necessary (it is three files you need to edit, config.js and two more for the colour scheme) and then open in the browser your local index.html.
No, there is no code that works without single cell data, maybe will come in the future.

**Comment by rocketeer1998** [comment]
For now, I can see my data has been successfully imported into the web page. But I'm stuck in the below snapshot for 10 minutes. Any recommendation?
![image](https://user-images.githubusercontent.com/55828283/174234668-ab1538e3-f3c0-409c-bb2c-dd261384c888.png)

**Comment by acycliq** [comment]
Most likely you missed to set the colours for you genes or cell classes. It's the `glyphConfig.js` you have to edit. Every single gene in your gene panel should be described there. If you miss one, the viewer will not load.
The same applies for the cell classes. Edit your `classColours.js` to set the colours for all your classes.

**Comment by rocketeer1998** [comment]
I've assigned each gene and each cell type a unique color, but the problem remains. All the three configuration files I used are uploaded to my repo. Could you help me check it out?

**Comment by acycliq** [comment]
Yes x1 and y1 are the width and height of your tiff in pixels

Can you send your browser log please? Typically in your Chrome in the dev-options of the browser (to go there press F12, or Ctrl+Shift+I or under the three vertical dots at the top right -> more tools -> developer tools)
Then select "Console" from the menu bar at the very top and send me a screen shot of that

**Comment by rocketeer1998** [comment]
![image](https://user-images.githubusercontent.com/55828283/174264631-8998048c-b1fb-421f-b90b-e9efa015c5d2.png)

**Comment by acycliq** [comment]
Yes probably one of your classes is missing from your classColors.js.  Add this line at the bottom

`{className: 'Zero', IdentifiedType: 'Zero', color: '#000000'},`

The `Zero` is the zero expressing class, if something doesnt fit the point pattern observed in your single cell data, it will be classified as Zero class. The Zero class doesnt exist in the single cell data, but if pciSeq cannot find a good match then it will be assigned to this "extra" class, the Zero class

**Comment by acycliq** [comment]
Great! Yes get back to me if you have any questions

**Comment by rocketeer1998** [comment]
Hi @acycliq , I'd like to create a gene-by-cell expression matrix based on pciSeq result. But I don't know how to assign each spot to each cell. Any recommendation?

**Comment by acycliq** [comment]
You can use the data from `cellData.tsv`, in particular the columns `Genenames `and `CellGeneCount`. The former (ie  `Genenames `) describes the genes that have been assigned to the cell and the latter (ie `CellGeneCount`) the corresponding gene counts. 

Is this what you need?

If you are wondering how come we have floats and not integers for the gene counts the answer is because this is a "soft clustering" method. It assigns the spot to a cell with some probability. For example lets say you had one spot and only two cells. Assume also, there was 50% prob for the spot to get assigned to the first and another 50% to the second cell. Then you would register 0.5 counts on each of the two cells. Under a hard clustering scheme where a spot is assigned to only cell, the you would register 1 count to one of the cells and 0 counts to the other.

**Comment by rocketeer1998** [comment]
That's what I want. How can I transform it into a matrix which has genes as rows and cells as columns? In addition, can I perform standard scRNA-seq data preprocessing, such as normalization and scaling, in this matrix? What's more, how can I export this .html file so that I can share it with other people without making a local server?

**Comment by acycliq** [comment]
Can you get the latest version, ` v:0.0.44`. Then if you call the `fit` method with `save_data` set to `True`, ie like 

`pciSeq.fit(_iss_spots, _coo, _scRNAseq, {'save_data': True})` 

it should also save a pickle file along the three `tsv` flatfiles. You can read the pickle by:
```
import pickle

infile = open(r'path\to\pciSeq.pickle', 'rb')
pciSeq_data = pickle.load(infile)
infile.close()
```

To get the array of the number of counts per cell you can do
`pciSeq_data.cells.geneCount` 
That will return an array of shape `Num_of_cells+1-by-genes`. The gene labels are sorted alphabetically but you can get them from `pciSeq_data.genes.gene_panel`

Also the top row of `pciSeq_data.cells.geneCount` corresponds to the background and it is filled with zeros, ie `pciSeq_data.cells.geneCount[0, :].sum()  = 0`  The background counts are kept elsewhere, here: `pciSeq_data.cells.background_counts`

Since you do not care about the background, to get the array you want do `pciSeq_data.cells.geneCount[1:, :]`.

Regarding your other questions, I do not know if you can use it to perform standard scRNA-seq data preprocessing.

For hosting the website somewhere else and not on your personal pc, If I have time I will write some notes duriing the weekend

**Comment by acycliq** [comment]
Can you send me your single cell data?

**Comment by rocketeer1998** [comment]
I've sent it to dimitris.nicoloutsopoulos@gmail.com.

**Comment by acycliq** [comment]
Thanks and sorry! Can you try with the latest build please, v 0.0.45. Let me know if you have any problems

**Comment by rocketeer1998** [comment]
Solved it, thank you! However, I find out the cell typing results are partially different before and after the release of `pciSeq(version=0.0.45)`. For example, a large number of cells previously annotated as TypeA are annotated as TypeB now. Are there any changes in the cell typing process of `pciSeq(version=0.0.45)`?

**Comment by acycliq** [comment]
Yes results will not be exactly the same but they should be quite similar where differences exist. Hopefully, in these cases, when you have differences results with the latest code will make more sense in general (or at least they wont be worse generally speaking)
